pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package statefulset
+
+import (
+	"sort"
+
+	"k8s.io/kubernetes/pkg/api/v1"
+	apps "k8s.io/kubernetes/pkg/apis/apps/v1beta1"
+
+	"github.com/golang/glog"
+)
+
+// StatefulSetControl implements the control logic for updating StatefulSets and their children Pods. It is implemented
+// as an interface to allow for extensions that provide different semantics. Currently, there is only one implementation.
+type StatefulSetControlInterface interface {
+	// UpdateStatefulSet implements the control logic for Pod creation, update, and deletion, and
+	// persistent volume creation, update, and deletion.
+	// If an implementation returns a non-nil error, the invocation will be retried using a rate-limited strategy.
+	// Implementors should sink any errors that they do not wish to trigger a retry, and they may feel free to
+	// exit exceptionally at any point provided they wish the update to be re-run at a later point in time.
+	UpdateStatefulSet(set *apps.StatefulSet, pods []*v1.Pod) error
+}
+
+// Returns a new instance of the default implementation StatefulSetControlInterface that implements the documented
+// semantics for StatefulSets. podControl is the PodControlInterface used to create, update, and delete Pods and to
+// create PersistentVolumeClaims. You should use an instance returned from NewRealStatefulPodControl() for any scenario
+// other than testing.
+func NewDefaultStatefulSetControl(podControl StatefulPodControlInterface) StatefulSetControlInterface {
+	return &defaultStatefulSetControl{podControl}
+}
+
+type defaultStatefulSetControl struct {
+	podControl StatefulPodControlInterface
+}
+
+func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, pods []*v1.Pod) error {
+	if set == nil {
+		return nilParameterError
```

Comment:
This is unusual - we don't guard against nil parameters in the majority of cases.